cmd/object-handlers.go

```diff
@@ -298,50 +285,57 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 
 	/// maximum Upload size for object in a single CopyObject operation.
 	if isMaxObjectSize(objInfo.Size) {
-		writeErrorResponse(w, r, ErrEntityTooLarge, objectSource)
-		return
-	}
-
-	// Size of object.
-	size := objInfo.Size
-
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		startOffset := int64(0) // Read the whole file.
-		// Get the object.
-		gErr := objectAPI.GetObject(sourceBucket, sourceObject, startOffset, size, pipeWriter)
-		if gErr != nil {
-			errorIf(gErr, "Unable to read an object.")
-			pipeWriter.CloseWithError(gErr)
+		writeErrorResponse(w, r, ErrEntityTooLarge, cpSrcPath)
+		return
+	}
+
+	cpMetadataOnly := strings.EqualFold(cpSrcPath, cpDestPath)
+	var newMetadata map[string]string
+	// Source and destination objects cannot be same if x-amz-metadata-directive is not set.
+	if cpMetadataOnly {
+		// Check if x-amz-metadata-direction was set to replace metadata.
+		if isMetadataReplace(r.Header) {
+			// Replace metadata requested proceed to extract
+			// relevant metadata from the request headers.
+			newMetadata = extractMetadataFromHeader(r.Header)
+
+			// Check if x-amz-metadata-direction was set to copy metadata.
+		} else if isMetadataCopy(r.Header) {
+			// Since copy was requested save what was previously saved.
+			newMetadata = objInfo.UserDefined
+
+			// Other headers like SSE(server side encryption)
+			// are explicitly not supported on Minio.
+		} else {
+			// For all other invalid values fail.
+			writeErrorResponse(w, r, ErrInvalidCopyDest, r.URL.Path)
 			return
 		}
-		pipeWriter.Close() // Close.
-	}()
-
-	// Save other metadata if available.
-	metadata := objInfo.UserDefined
-
-	// Remove the etag from source metadata because if it was uploaded as a multipart object
-	// then its ETag will not be MD5sum of the object.
-	delete(metadata, "md5Sum")
+	}
 
-	sha256sum := ""
+	// Hold write locks on destination if metadata is not to be updated
+	// on source but instead destination object is being created.
+	objectDWLock := globalNSMutex.NewNSLock(dstBucket, dstObject)
+	objectDWLock.Lock()
+	defer objectDWLock.Unlock()
 
-	objectWLock := globalNSMutex.NewNSLock(bucket, object)
-	objectWLock.Lock()
-	defer objectWLock.Unlock()
+	// Hold read locks on source object only if we are going to read
+	// data from source object. For metadata updates source and destination
+	// object will be same so holding write locks on destination is good
+	// enough to safely attempt concurrent access.
+	if !cpMetadataOnly {
+		objectSRLock := globalNSMutex.NewNSLock(srcBucket, srcObject)
+		objectSRLock.RLock()
+		defer objectSRLock.RUnlock()
+	}
 
-	// Create the object.
-	objInfo, err = objectAPI.PutObject(bucket, object, size, pipeReader, metadata, sha256sum)
+	// Copy source object to destination, if source and destination
+	// object is same then only metadata is updated.
+	objInfo, err = objectAPI.CopyObject(srcBucket, srcObject, dstBucket, dstObject, newMetadata)
```

Comment:
to simplify locking logic, this can be done at the beginning:
if (src == dst) then hold write lock on src
else hold read lock on src and write lock on dst